File: src/2019/AdventOfCode2019/Day01.cs
using AdventOfCodeLib;

namespace AdventOfCode2019;

public sealed class Day01 : IAocDay<int>
{
    public static int Part1(AocInput input) =>
        input.Lines
            .Select(int.Parse)
            .Sum(Fuel);

    public static int Part2(AocInput input) =>
        input.Lines
            .Select(int.Parse)
            .Sum(TotalFuel);

    private static int Fuel(int mass) => mass / 3 - 2; // => (int)(Math.Floor(mass / 3.0) - 2);

    private static IEnumerable<int> FuelRequirements(int mass)
    {
        var fuel = Fuel(mass);
        while (true)
        {
            if (fuel < 0)
            {
                yield break;
            }

            yield return fuel;
            fuel = Fuel(fuel);
        }
    }

    private static int TotalFuel(int mass)
    {
        var fuel = Fuel(mass);
        return fuel < 0 ? 0 : fuel + TotalFuel(fuel);
    }
}

File: python/y2019/day01.py
from aoc.core import solve, AocInput
from functools import cache


def fuel(mass: int) -> int:
    return mass // 3 - 2


@cache
def totalFuel(mass: int) -> int:
    f = fuel(mass)
    return 0 if f < 0 else f + totalFuel(f)


def part1(input: AocInput) -> int:
    return sum(map(fuel, map(int, input.lines)))


def part2(input: AocInput) -> int:
    return sum(map(totalFuel, map(int, input.lines)))


def main() -> None:
    YEAR = "y2019"
    DAY = "Day01"
    print(f"Running year {YEAR.replace('y', '')} day {DAY.replace('Day', '')}")
    solve(YEAR, DAY, part1)
    solve(YEAR, DAY, part2)
